drivers/power_domain/power_domain_nrfs_gdpwr.c

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrfs_gdpwr
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/logging/log.h>
+
+#include <nrfs_gdpwr.h>
+#include <nrfs_backend_ipc_service.h>
+
+LOG_MODULE_REGISTER(nrfs_gdpwr, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+#define MANAGER_REQUEST_TIMEOUT K_MSEC(CONFIG_POWER_DOMAIN_NRFS_GDPWR_TIMEOUT_MS)
+
+static K_SEM_DEFINE(lock_sem, 1, 1);
+static K_SEM_DEFINE(req_sem, 0, 1);
+static nrfs_gdpwr_evt_type_t req_resp;
+static const struct device *const domains[] = {
+	DT_INST_FOREACH_CHILD_SEP(0, DEVICE_DT_GET, (,))
+};
+
+struct domain_data {
+	bool off;
+	bool synced;
+};
+
+struct domain_config {
+	gdpwr_power_domain_t domain;
+};
+
+static void manager_event_handler(nrfs_gdpwr_evt_t const *evt, void *context)
+{
+	ARG_UNUSED(context);
+
+	req_resp = evt->type;
+	k_sem_give(&req_sem);
+}
+
+static void manager_lock(void)
+{
+	if (k_is_pre_kernel()) {
+		return;
+	}
+
+	(void)k_sem_take(&lock_sem, K_FOREVER);
+}
+
+static void manager_unlock(void)
+{
+	if (k_is_pre_kernel()) {
+		return;
+	}
+
+	k_sem_give(&lock_sem);
+}
+
+static int manager_set_domain_locked(gdpwr_power_domain_t domain, bool on)
+{
+	nrfs_err_t err;
+	gdpwr_request_type_t req = on ? GDPWR_POWER_REQUEST_SET : GDPWR_POWER_REQUEST_CLEAR;
+	int ret;
+
+	err = nrfs_gdpwr_power_request(domain, req, NULL);
+	if (err != NRFS_SUCCESS) {
+		LOG_ERR("%s %s", "nrfs gdpwr request", "failed");
+		return -EIO;
+	}
+
+	ret = k_sem_take(&req_sem, MANAGER_REQUEST_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("%s %s", "nrfs gdpwr request", "timed out");
+		return -ETIMEDOUT;
+	}
+
+	if (req_resp != NRFS_GDPWR_REQ_APPLIED) {
+		LOG_ERR("%s %s", "nrfs gdpwr request", "rejected");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int manager_set_domain(const struct device *dev, bool on)
+{
+	struct domain_data *dev_data = dev->data;
+	const struct domain_config *dev_config = dev->config;
+	int ret;
+
+	manager_lock();
+
+	if (dev_data->synced) {
+		ret = manager_set_domain_locked(dev_config->domain, on);
+	} else {
+		ret = 0;
+		dev_data->off = !on;
+	}
+
+	manager_unlock();
+	return ret;
+}
+
+static int manager_sync_domain_locked(const struct device *dev)
+{
+	struct domain_data *dev_data = dev->data;
+	const struct domain_config *dev_config = dev->config;
+
+	dev_data->synced = true;
+
+	if (dev_data->off) {
```

Comment:
this expects domains to be by default requested after reset right? Can you leave a comment about this, otherwise it seems strange why you wouldn't just do unconditional `manager_set_domain_locked(dev_config->domain, !dev_data->off);`

Reply:
Yep, will leave a note about this :)